torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2983,24 +2989,23 @@ def _pre_backward_hook(_handles: List[FlatParamHandle], *unused: Any) -> None:
                 for handle in _handles:
                     handle._training_state = HandleTrainingState.PRE_BACKWARD
 
-                params_prefetched = self._handles_prefetched.get(_handles_key, False)
-                if not params_prefetched:
-                    self._unshard(_handles)
+                # If the handles have been prefetched, this `_unshard()` simply
+                # switches to using the unsharded parameter
+                self._unshard(_handles)
```

Comment:
Having the `if not params_prefetched` led to a subtle bug with activation checkpointing where a handle would be prefetched and then activation recomputation would trigger, resharding that handle without marking that the handle as no longer prefetched.

To fix this, we adjusted some logic, but importantly, because we have written `_unshard()` to avoid unnecessary allocations, it is fine to freely call it here without the `if params_prefetched` guard.